### ISSUE TITLE
Request for deletion

### ISSUE THREAD

**Issue by steipete** [ghtorrent/ghtorrent.org#32]
Hi,

I'm getting a ton of emails for surveys/research around software, and it's really annoying. I recognize the scientific effort and initially helped many people here, however after so many requests, I feel that I did my part and would like to opt out. Can you remove all of my data/emails from this giant data blob?

Thank you.

**Comment by futuretap** [comment]
Please remove my data, too.

**Comment by slang800** [comment]
:-1: - There are a few reasons why this is a bad idea:

- Removal from the GHTorrent dataset won't get rid of your data. Both of your email addresses are still publicly available on GitHub. @steipete has his in his profile (see below), and while @futuretap doesn't, it's still in every single commit (here's the [first commit I could find](https://github.com/futuretap/InAppSettingsKit/commit/df2e52780a5abd7b3af102adb01ef9752f43f408.patch), for example).

  ![1456479469](https://cloud.githubusercontent.com/assets/1049204/13349045/4d7a0b74-dc45-11e5-8df1-5330d967182c.png)


- GHTorrent is ~6TB of data right now, which would all need to be processed and re-released. Not only is that kind of processing impractical for ad-hoc requests, but everyone who has already downloaded the datasets would need to be found (somehow), contacted, and asked to redownload the cleaned versions.
- Removing people would introduce errors into the data-set - like issues or PRs created by non-existent people. These kinds of errors would likely prompt researchers and archivists to explicitly seek out unaltered data through other means (see next point).
- GHTorrent isn't the only dataset, it's just a centralized copy. Anyone can run their own copy of GHTorrent and make their own dataset. In fact, it's not even the only project - [githubarchive](https://www.githubarchive.org/) probably has a copy of your data too and I'm sure that [archive.org](http://archive.org) has a copy somewhere.

So, I think you'd have much better luck just switching to a throwaway email for your GitHub accounts & git commits. Also, it might be a good idea to establish a list of people who do not want to participate in GitHub-related research projects. Complying with such a list would obviously be voluntary, but I don't think that many researchers want to annoy. So, if ghtorrent.org added a note about the existence of a list to important locations like the downloads page, I bet it would be more effective at solving this than data-removal.

**Comment by steipete** [comment]
It's not about removing my email address completely - it's all about making it less *convenient* for people to get a complete set of scraped data, that they can use to send everyone spam.

My email is public and that's fine, but there's a difference between people browsing my profile and deciding to ping me and people sending emails to tens of thousands of emails because they have one convenient data set and it doesn't cross their minds that other people could be offended by the sheer amount of spam emails this creates. The problem is that it's not something GMail can filter automatically since these are legitimate emails - just way too many.

**Comment by futuretap** [comment]
Yes, it's all about making it easy or not for spammers. Your database just makes it way too easy. So please remove my address. You may well replace it by something like no@address if you care about consistency.

**Comment by gousiosg** [comment]
[Apologies for the late reply, I 've just seen this]

@steipete I realize that this may be inconvenient, I am also getting quite a few of those emails as well. However, "opting out" is not really something realistic; as @slang800 also writes, this is public data that GHTorrent just collects; Google, Bing, GitHub Archive, GitHub's own search and various other sites that link to Github also do the same and you can readily query them for developer profiles.

Personally, as a researcher, I think that having access to such a vast pool of interconnected data is amazing. As an individual, I don't like much that other people are building a profile of me by judging my OSS activity. However, I "opted-in" when I shared my email and agreed to GitHub's terms, so I should have been more careful.

I like @slang800's suggestion and therefore I will create a page at GHTorrent (top-level) so that people who do not agree with being contacted about research can be listed there. It is however up to the researchers to comply. As far as I understand research, most researchers will be happy to comply.

**Comment by gousiosg** [comment]
@futuretap I would hardly call researchers contacting developers for input "spammers".

**Comment by futuretap** [comment]
Honestly, that opt-out page is an even worse idea. Spammers will love this list even more so it exposes my address even more.

The difference to Google and Bing is you can't ask those search engines: "Give me a list of thousands of email addresses to spam to". So I stand by my opinion that exposing email addresses like this is a terrible idea. Why don't you remove just the email column? Interesting researchers could still find out the addresses manually but you wouldn't create a spammer's heaven database.

**Comment by futuretap** [comment]
@gousiosg how can you be sure that only researchers use the data? In fact I'm pretty sure, most abuse is not being done by researchers.

**Comment by gousiosg** [comment]
@futuretap I cannot remove the email column because it identifies devs that do not have a GitHub account and links them with their commits. 

WRT the white list, we can have the white listed people's email replaced by something like 'no-spam@ghtorrent.org'. I do need to have a whitelist so that GHTorrent actually consults it when it is processing updates.

**Comment by gousiosg** [comment]
@futuretap WRT your comment on search engine usability

<img width="764" alt="screen shot 2016-02-26 at 12 26 44" src="https://cloud.githubusercontent.com/assets/386172/13351039/43a54098-dc84-11e5-8f32-151d8ed9d80c.png">

```
curl 'https://api.github.com/search/users?q=language%3AObjective-C%20followers%3A%3E%3D150&order=asc&sort=followers&per_page=100&page=1'|grep futureap
```

**Comment by futuretap** [comment]
I'm a bit confused. You're calling this opt-out list a "white list"? I'd call it a black list.

Anyway, the point is, when opting out, I'd like to be exposed **less** not more with my email address.

So do I understand correctly that when opting out, you'll replace the email in the database with something like 'no-spam@ghtorrent.org'? Then I'm all in!

WRT your search engine comment, I don't understand your message here. This search thankfully doesn't  include email addresses. But your database does.

**Comment by slang800** [comment]
> The difference to Google and Bing is you can't ask those search engines: "Give me a list of thousands of email addresses to spam to". - @futuretap 

Au contrair, mon ami - that's exactly how spammers build initial lists. Check out http://www.binarytides.com/email-harvesting-metasploit/ for a tutorial.

**Comment by futuretap** [comment]
I think it's moot to discuss spammer techniques here. I still think it's reasonable to ask you to remove my address from your database.

**Comment by gousiosg** [comment]
@futuretap Ok, for the white/black list, I am glad you agree!

WRT to GitHub's search, let me elaborate a bit:

```bash
curl 'https://api.github.com/search/users?q=language%3AObjective-C%20followers%3A%3E%3D150&order=asc&sort=followers&per_page=100&page=1'|
grep '"url": '|sed -e 's/"url": "\(.*\)",$/\1/'|
while read url; do 
    curl -s $url
done |
grep email
```

With a little more effort (and a few more API keys) I can collect thousands of emails in a couple of hours. My point is that it is pointless to remove your email from GHTorrent, but since you insist, I have done this already (also for @steipete). You can confirm this here: http://ghtorrent.org/dblite/

```sql
​select * from users where login='futuretap'​​​
```

**Comment by milosonator** [comment]
You can do that but it will be extremely ineffective, the spammers clearly already have it, and asking to remove it is like asking to remove your phone number from the new phone-books when your phone number is already all over the internet, ready to be harvested by anyone who can write some bash, as @gousiosg displays here.

If you want your e-mail to be more private, remove it from GitHub altogether to prevent future spammer to grab it.

**Comment by slang800** [comment]
> I'm pretty sure, most abuse is not being done by researchers. - @futuretap 

What makes you think that?

When I initially replied I assumed that you were talking about emails from researchers that explicitly say they found your info from GHTorrent. But if you're assuming that spammers have been finding your info from this dataset, then I think you should consider how much easier it is to just scrape GitHub itself and avoid the huge download... Or maybe just have Google scrape it for you, since it seems like they cache the email address too:

![1456487733](https://cloud.githubusercontent.com/assets/1049204/13351704/9d9c0c22-dc56-11e5-9d2c-75a6dfe8433c.png)


> I think it's moot to discuss spammer techniques here. - @futuretap 

Perhaps, but my point is that GHTorrent isn't unique in this problem... It sucks if spammers are using this dataset, but there isn't a decent solution... not even search engines have found a good way to protect emails from spammer harvesting.

**Comment by samdmarshall** [comment]
I second that @futuretap, there should be a way to opt out of this service. The amount of "oh I found you on Github" Mail I've been getting lately is incredibly annoying.

**Comment by gousiosg** [comment]
Researchers have been contacting devs before GHTorrent; in fact half of the research papers being published mention Github Archive as a source of dev data. As I demonstrate above, it is easier to just scrap GitHub's search than download and restore (takes 2-3 days!) GHTorrent's database to collect emails (not to mention the `/users` API endpoint).

It is 2016; if you expose your email on the Internet, you are bound to get emails (spam or legitimate ones).

**Comment by steipete** [comment]
It's really not classial spammers but students and researchers who are completely oblivious to the problem and consider the existance of this project als legitimicy to email=spam everyone. 

The last time I tried to explain the issue, the researcher suggested to sending an email to everyone, asking if they are getting too many emails from this source. By then I just wanted to scream.

**Comment by futuretap** [comment]
Thanks @gousiosg for the removal. If you're serious about fighting spam, you might want to reconsider publishing the harvesting script here. I get your point that it's easily doable also without you publishing the detailed spammer harvesting recipe.

**Comment by gousiosg** [comment]
@futuretap If I can do it in 5 mins, one has to assume that any developer can do it. Security (or privacy) through obscurity does not work.

**Comment by futuretap** [comment]
As @steipete said, it's more about dumb students and researchers than professional spammers.

**Comment by steipete** [comment]
I opened this issue with the hope that it might cause a change in mindset. Just because it's easy to ring on all doors on the street doesn't mean you should do it.

**Comment by slang800** [comment]
> it's more about dumb students and researchers than professional spammers

Then shouldn't we try to educate them, rather than punish _everyone_ with an inferior dataset?

**Comment by futuretap** [comment]
Absolutely. However that education shouldn't include spam email harvesting techniques.

**Comment by steipete** [comment]
Yes. You could educate them best by force-pushing this repo to an empty one with a readme explaining the problem. Otherwise I doubt people will read.

**Comment by gousiosg** [comment]
Oh right, because censorship always solves problems with 'dumb' people using information, doesn't it?

**Comment by madrobby** [comment]
@gousiosg Please remove me as well. I've never agreed to share my data on your website. Thank you.

**Comment by steipete** [comment]
How about adding a disclaimer on the start page of http://ghtorrent.org explaining the issue and asking people to not send unsolicited mass-email. Currently it does not mention anything on this highly relevant topic there.

Some people might believe it's common sense to not send email to a huge list without an explicit opt-in, but apparantly this isn't the case if one can justify it under the research-label.

**Comment by mallamanis** [comment]
GHTorrent is a valuable resource for software engineering research and I disagree with any idea of censoring it. 

People that may be 'dumb' to send mass emails are not necessarily "technically dumb" to _not_ be able to retrieve emails in another way. Removing any data from GHTorrent will not change the way some people (researchers or not) behave and I feel that it is *not* GHTorrent's problem to solve. Developing in public implies that all the commits are visible to everyone and by definition there is no privacy in that. If your email has gone public, it *is* public (and this is your problem). [just to note here: this is the case with my email too. I don't like it, but I accept it]

Philosophical question: @steipete How do you even know that the unsolicited emails you have received from researchers/students were extracted from GHTorrent? For a long time I was using GitHub's own dump in Google BigQuery with [GitHub Archive](https://www.githubarchive.org/) that also contains this data and it should be equally easy to get an email dump from there. I have also received multiple emails because of people that have scraped app store emails. Removing the data from GHTorrent definitely will not solve the problem with email etiquette.

Some suggestions:

- (as @steipete suggested) A disclaimer: the emails in the dataset is _not_ a mailing list. Researchers should not use the dataset as a mailing list. Individuals _cannot_ unsubscribe from it (because it's not a mailing list).
- A FAQ: People need to understand that once they use their email publicly (e.g. in a git commit or in GitHub elsewhere) their email might, may and will be found by anyone. GHTorrent or anyone else cannot stop that. The FAQ could explain how to avoid getting your email to appear in public when using git/GitHub.

**Comment by steipete** [comment]
I asked the last researcher who sent mass-email where he has the data from, and he mentioned this archive. That's how I found out about this in the first place. This doesn't mean all research-email-spam (I keep calling unsolicited mass-email spam, even if it is for science) is based from this data set though, so my evidence is limited. But since I get quite a few emails that are all based on the same scheme, I can ask the next few ones as well.

**Comment by sbaltes** [comment]
> The last time I tried to explain the issue, the researcher suggested to sending an email to everyone, asking if they are getting too many emails from this source. By then I just wanted to scream.

@steipete This is not true. We discussed about building awareness in the research community that many developers are annoyed by the amount of survey invitations they receive. I wrote that one needs to be able to quantify the problem to build awareness. And that was my suggestion how one could quantify the problem and try to publish the results on a large academic conference:

> 1. Conducting a survey with researchers that published results of a developer survey recently on one of the major academic SE conferences like ICSE and FSE, asking them how they did their sampling and if and how they cared about ethical issues (like annoying developers).
> 
> 2. Somehow getting a response from 100-150 developers indicating the amount of "survey spam" they receive and a short comment on the problem. One option would be to use the GHTorrent data source again, or trying to reach them e.g. over Twitter. 

I did not say anything about contacting everyone, I talked about sampling some devs using GHTorrent or Twitter.

BTW: For my last survey, I only contacted GitHub users who had their email address publicly available on their profile page.

To the researchers reading this: I'm still planning to conduct the first survey (asking researchers about ethical considerations). If anyone is interested, just contact me.

**Comment by ashfurrow** [comment]
> BTW: For my last survey, I only contacted GitHub users who had their email address publicly available on their profile page.

My email is only there so people can contact me privately about problems in the community. You're assumption that public email = feel free to contact me about stuff that interests _you_ is unfounded.

**Comment by slang800** [comment]
> My email is only there so people can contact me privately about problems in the community.

We should probably compile a opt-out list of people for whom that is the case, because my email is there for a totally different reason (I'm happy to be contacted by researchers), and yet neither of our profiles make that distinction.

**Comment by mallamanis** [comment]
@ashfurrow it's interesting you raise this point. There are two faces to what you are describing:

a) Practitioners do not get/believe/understand most (or all?) of the software engineering research and its results. Or they don't care about the topics that researchers are working on. If this were the case, the "problems in the community" would be exactly the stuff that researchers would care about. 

b) On the other hand @ashfurrow, how can researchers learn about "the problems in the community" if they cannot contact you? Unless, of course, you believe that research/researchers cannot help you with those problems.

**Comment by ashfurrow** [comment]
You misunderstand. I'm not referring to general problems in the community, I mean I want people being bullied/harassed/etc on any project I'm involved with to have a way to privately contact me.

**Comment by neilernst** [comment]
The time and energy people have spent complaining that the public email they published on Github is being used for a quasi-legit purpose could have been better spent creating a simple filter (or just clicking delete, honestly).

@gousiosg My suggestion is to include on the GHTorrent page a reminder that emails included in the dataset are not agreement to be contacted. And maybe some terms of use are needed?

**Comment by ashfurrow** [comment]
We can't create a filter, as @steipete pointed out.

And you don't get do tell us how to spend our time and energy.

**Comment by danielr** [comment]
> I cannot remove the email column because it identifies devs that do not have a GitHub account and links them with their commits. (@gousiosg)

If identification is the main reason here, why don't you just hash them?

**Comment by gousiosg** [comment]
@danielr This can be an option, but it only adds an extra step to the process of retrieving emails. You can always query dev actions, find a few thousand dev usernames and get their emails through GitHub's API. Determined users (and I can assure you that researchers are pretty determined users) will surely walk the extra mile.

I will repeat it one final time: It is pointless to try and conceal information that is otherwise open on the Internet.

**Comment by futuretap** [comment]
Hashing emails would be a sensitive solution imo. With offering a service comes responsibility so an additional step for that seems more than appropriate.

**Comment by slang800** [comment]
I agree with @gousiosg on the pointlessness of adding extra steps... we should all be familiar with what happens when you try to [hide something on the internet](https://en.wikipedia.org/wiki/Streisand_effect). If GHTorrent doesn't provide emails in an easy format then someone else will inevitably step up and provide them in its place.

**Comment by milosonator** [comment]
I like the hashing idea.

2016-02-26 16:05 GMT+01:00 Sean Lang <notifications@github.com>:

**Comment by danielr** [comment]
@gousiosg I think nobody here argues that you are in the position to make it impossible to find publicly available addresses (and I don't think that this was @steipete's point). But you can make it *harder* to *harvest* those addresses via *your* dataset. Honestly, I don't see any reason not to hash them.

**Comment by gousiosg** [comment]
I will not hash emails because i) it is extra effort for minimal (if any) benefit and ii) will deteriorate the quality of research being done with GHTorrent. People are using emails to link GitHub profiles to StackOverflow, OSS project mailing lists and Jira databases and perhaps other data sources as well. You can have a look at the proceedings of the [MSR](http://2016.msrconf.org/) to get an idea what emails are used for.

**Comment by slang800** [comment]
> Honestly, I don't see any reason not to hash them.

You don't see _any_ reason? How about the burden on researchers who will have to go download another dataset and join it with the column of email hashes, just because someone thought that security through obscurity works? Or the inconsistency in the format of releases that this change would cause. Or all the scripts that rely upon GHTorrent having an email column.

**Comment by futuretap** [comment]
Maybe @github should step up and change their terms so that mining emails is not allowed anymore if not opted-in by the user.

**Comment by samsonjs** [comment]
In many countries the law saws you must 1) tell people what personal info you have about them, and 2) remove it at their request.

How on earth are either of those unreasonable requests?

**Comment by slang800** [comment]
@futuretap: that would be pointless. You publish your email by using git... Every single commit is signed with it.

And how do you think GitHub would enforce such a rule? If you're worried about spammers, then you're not dealing with people who care about the terms of service.

**Comment by mallamanis** [comment]
@samsonjs @slang800 I feel very protective of my email, but hardly anyone would argue that a user's email addresses is personal information. Especially when the user has publicly and knowingly disclosed in public this information (through GitHub and/or any public git repository). 

A relevant (but maybe too controversial) article about a similar issue: [Your Email Address Isn't Personal Information. Get Over It.](https://newrepublic.com/article/116211/gmail-feature-raises-privacy-concerns-about-email)

**Comment by futuretap** [comment]
@slang800 This would solve that GHTorrent and others can't harvest email addresses anymore. There's a difference in publishing email addresses on individual pages (for instance, I do publish email addresses on my company website) and mass-harvesting and **mass-distributing** them.

**Comment by SpacyRicochet** [comment]
@slang800 
> If you're worried about _spammers_, then you're not dealing with people who care about the _terms of service_.

**Github ToS:**
> G.10. You must not upload, post, host, or ** --> transmit unsolicited email <--**, SMSs, or "spam" messages. 
(emphasis mine)

Apparently researchers exhibit the same behavioural patterns as spammers.

I wasn't going to comment originally, but two things changed my mind;

1. The "researchers" complete and utter **lack of concern for people's privacy**. Yup, anyone that argued that removing someone's personal information from a _publicly accessible database they own_ at the person's request is somehow useless/difficult/against science doesn't care about privacy and apparently has questionable ethics… period.
2. This project has apparently originated in The Netherlands. **We have explicit laws against this behaviour!**

(╯°□°）╯︵ ┻━┻

**Comment by slang800** [comment]
@futuretap - what is that difference? Would you be guilty of mass-distribution just for mirroring or archiving GitHub? Or does the problem arise when you provide data in an easy-to-access format? Could you get banned just for posting the list of emails in the `git log`?

**Comment by samsonjs** [comment]
@mallamanis @slang800 Email sent to my address comes to my inbox. It's 100% personal. You let me worry about the other places *I choose* to distribute it. I did not choose to distribute it in this massive list.

**Comment by slang800** [comment]
@SpacyRicochet - GHTorrent isn't a researcher, doesn't control any researchers, and isn't sending anyone emails. It's a data-set. I think that we all agree that spamming is bad. The debate that we're having is about whether or not censoring GHTorrent is ethical and/or useful.

I'm actually not familiar with Dutch legal code, are you talking about the "right to be forgotten", or something else?

**Comment by neonichu** [comment]
Let's also be honest here, the main contributor of GHTorrent **is** a researcher and is admitting to spamming people using GHTorrent data on their blog: <http://www.gousios.gr/blog/Scaling-qualitative-research/>

**Comment by futuretap** [comment]
Interesting. So @gousiosg is actually violating the @github ToS. If not by distributing the database then by sending those 3,500 emails as he openly admits.

**Comment by slang800** [comment]
@samsonjs / @SpacyRicochet: You both _did_ distribute your email addresses. I truly don't understand how you can publish this information on the internet, and then be upset when someone makes it too easy to access.

If I argued that my quote saying "spammers don't care about the ToS" was private information and @SpacyRicochet was infringing upon my rights & redistributing it without my consent by reproducing it in his [comment](https://github.com/ghtorrent/ghtorrent.org/issues/32#issuecomment-189348271), would you not think I was insane?

**Comment by segiddins** [comment]
My email _needs_ to be public, for the same reasons Ash pointed out -- largely, people need to be able to contact me with CoC-related issues for the projects that I run. I don't mind my email being public, but I _do_ mind it being aggregated in this way, and am frankly quite bothered by the amount of spam I receive, and the fact that _you refuse to remove my personal data from a datastore you control and distribute publicly_.

**Comment by samsonjs** [comment]
@slang800 I didn

**Comment by samsonjs** [comment]
@slang800 I didn't consent to or personally distribute my email in this giant torrent. I disclose it on GitHub. If a spammer scrapes GitHub and emails me then that's a separate issue that has no bearing on this discussion. You don't get to tell me that because I distribute my info in one way then others suddenly have the right to take my info and distribute it in other ways.

**Comment by samsonjs** [comment]
@slang800 How about I go to your website and copy posts from your blog. I mean, you published it on the Internet so now it's fair game for anyone to use for any purpose they wish, right? Give me a break.

**Comment by slang800** [comment]
@samsonjs: if they're CC0'd, then feel free to copy verbatim. Otherwise just give attribution and you're good.

**Comment by mallamanis** [comment]
I believe that you are using GHTorrent as a scapegoat for the spam you are receiving by researchers that do not do research "ethically". We all agree that spam is annoying. We shouldn't be discussing this here. Instead we should be discussing if removing the emails from this dataset (that contains information that is already publicly available to everyone with an Internet connection) will stop those emails from coming (which are indeed very annoying).

As @segiddins and @ashfurrow  say the email address is there to _solicit_ emails from a specific subset of people. Unfortunately, no one can enforce a subset on this solicitation with current GitHub's structure. It's not GHTorrent's fault. @github should allow us to set specific permissions to allow disclosing our email (e.g.) only to contributors.

Regarding your previous post: @samsonjs The emails that arrive at your inbox are 100% personal and private. No one else can/should read them or even know that you have received any email or know about any of its metadata. But an email address is not private information once publicly disclosed. (We both chose to publicly disclose this information at some point). I hate getting spam (from researchers or not) but I have to live with it, as I have to live with the unsolicited physical ad brochures I receive at my office. Both these instances of spam do _not_ violate anyone's privacy (but are very annoying).

So, if you buy my argument above, please, lets stop talking about privacy and focus on spamming. And in specific if removing this information from here, will reduce our spam influx (I believe this isn't the case).

Enjoy your weekend :)

**Comment by lazerwalker** [comment]
This is more a case of respect rather than literal privacy. We all recognize and understand this information is already public on the Internet, and you can technically do what you want with it. Many of us  believe that having our email addresses and other data removed from your dataset would decrease the amount of spam. Maybe we're right, maybe we're wrong; what's more important is that you refusing to do this shows an incredible lack of respect and empathy for the people whose data you're scraping and making more readily available.

The Internet Archive's Wayback Machine is a useful comparison, as it's a similar use case: publicly-available information is being captured and exposed in a way that makes it available even after the original information is gone, and in some cases may make it easier to find. 

The Wayback Machine lets you opt-out. They very clearly tell you exactly how to stop them from scraping you, either by modifying your robots.txt file or just by emailing them (https://archive.org/about/faqs.php#2). Notably, they retroactively this; if you add a robots.txt file to your site today blocking the Internet Archive, any former versions of your site that showed up on the Wayback Machine will be hidden going forward.

Even though their goal is to archive the entire Internet, they recognize that respecting Internet users enough to give them the choice to opt out is more valuable than having an exhaustive data set or simplifying their maintenance.

**Comment by alloy** [comment]
I would like to +1 Mike's first paragraph, which is exactly how I feel about it. As such, please add me to the list of people not to contact.

**Comment by ELLIOTTCABLE** [comment]
So, I usually don't want to contribute to a pile-on like this, but this is *so* *crucial*:

> I will not hash emails because i) it is extra effort for minimal (if any) benefit and ii) will deteriorate the quality of research being done with GHTorrent. People are using emails to link GitHub profiles to StackOverflow, OSS project mailing lists and Jira databases and perhaps other data sources as well. You can have a look at the proceedings of the MSR conference to get an idea what emails are used for.

No. Stop. *You* don't get to say how much benefit it is to *us*: your argument that it's “just as easy to get the information elsewhere” isn't fucking enough. We *know* it's technologically trivial, that's not the point; that's like arguing that “It's okay that I'm forging hundreds of signatures, because it's really easy to forge a signature! Anybody could do it!”

*Obscuring* the e-mail address in the database (preferably, via a simple hash) allows the researchers to do the same matching of username-to-commit that is currently done via e-mail address [which is a valid goal]; but, and let me put this in large type, because it's *really* *important*, …

#### … doing so sends an important message.

It says to the (possibly not-so-concerned-with-us-developers'-time-and-energy) researchers using your database, “hey, if I need to take an extra step to scrape all these e-mail addresses … huh, this, kinda feels like the kind of thing that I hear about spammers doing … wait, does this make me a spammer?” (Hint, researchers reading this: yes. yes it does. You are a spammer. Please stop.)

----

And, more importantly than refuting your “there's no point to you doing this” argument, there's this one:

#### This is our data. Not yours.

From your point of view, we may have posted this publicly, in one fora (on a social-networking site, *for the use of other developers* wishing to contact us about the software work we're explicitly doing *with* those other developers*), but that *does not imply an opt-in to such research efforts.*

This sort of invasion of privacy? It's why IRBs exist; it's why research subjects are guaranteed full-disclosure and informed consent is sought.

----

I seriously cannot believe you're arguing about this, if you call yourself a researcher. Where are your scientific ethics? This is a *huge* privacy, disclosure, and informed consent issue. I'm sorry if this post seems adversarial, but that's because *you've hurt me*, and *you have no right to*. Get your head on straight, and remember the ethical foundations of your field as a researcher, please.

**Comment by davidschreiber** [comment]
+1 @ELIOTTCABLE

**Comment by gousiosg** [comment]
Let me clarify a few things:

* I have never refused removing individuals' emails from the GHTorrent database. If you bothered to read the whole thread you would see that I have already done so. I also proposed a scheme where individuals would be able to have their emails removed (@slang800 has already started working on it).
* If something is on the Internet, you either need to provide a license with it or it is in the public domain. This is why licenses such as Creative Commons where invented.
* To the best of my knowledge, I have not used the GitHub service to "upload, post, host, or transmit unsolicited email" for my research.
* I do not allow anyone question my scientific integrity, scientific ethics or my research work on grounds other than scientific merit. My research work is fully disclosed on this very site and you are welcome to inspect it. As my 1,500 respondents can confirm, I 've treated their privacy with the utmost respect.

**Comment by segiddins** [comment]
> If something is on the Internet, you either need to provide a license with it or it is in the public domain. This is why licenses such as Creative Commons where invented.

False. By default, at least in the USA, _fully copyright is the default_. No license may be presumed in the absence of a grant thereof, and public domain _is_ a license.

> I have never refused removing individuals' emails from the GHTorrent database

In that case, please remove mine as well.

**Comment by gousiosg** [comment]
@segiddins I did not discuss copyright, which depending on the jurisdiction may or may not, belong to you. I discussed licensing. If something is public domain, then derivative work is allowed by default, at least according to my understanding of IP law (IANAL).

**Comment by davidcelis** [comment]
Please remove me as well. Thanks.

**Comment by chucker** [comment]
> I did not discuss copyright, which depending on the jurisdiction may or may not, belong to you. I discussed licensing.

Licensing is only relevant *because* of copyright. If a work isn't copyrighted, it doesn't need to be licensed, as it's effectively in the public domain. These claims of yours:

> If something is on the Internet, you either need to provide a license with it or it is in the public domain. This is why licenses such as Creative Commons where invented.

…are way off, and dangerous for a project that so significantly hinges on third parties' rights — privacy, property or otherwise.

**Comment by cainlevy** [comment]
There's a world of difference between technically available information and readily available information. The lock on my front door is technically not secure, but in practice it still keeps out the majority of undesired visitors.

This data set makes it easy to infer private personal details that individuals would prefer remain difficult to discover. Please anonymize all research data sets. I would suggest dropping most of the users table, except for maybe `country_code`. Yes, it reduces the potential uses for the data set but that is actually the point.

Meanwhile, please remove me.

**Comment by rastersize** [comment]
**Please remove any data about me and block my personal data from being included again.** Also make sure such blocking is done in a privacy friendly manner.

Why, really because it’s the right thing to do morally and ethically. Respecting me, my personal data and how I want it to be handled. However if that’s not enough I’ll ask you to read the “So why should you comply with my wish of removal?” which details the legal aspects.

I didn’t know about this data collection and processing until I was alerted via a random tweet and looked up myself. Finding out that my personal data is being collected and processed. **You do not have my consent** to collection and processing my personal data. Not cool.

### So why should you comply with my wish of removal?
It boils down to personal data protection laws in EU as defined by [Directive 95/46/EC](http://eur-lex.europa.eu/LexUriServ/LexUriServ.do?uri=CELEX:31995L0046:en:HTML) of the European Parliament and of the Council. The Netherlands (which GHTorrent is based in according to your public information) [has ratified this directive](http://ec.europa.eu/justice/data-protection/law/status-implementation/index_en.htm#maincontentSec16) and passed it as “[Wet bescherming persoonsgegevens](http://wetten.overheid.nl/BWBR0011468/2010-07-01)” (Personal Data Protection Act) which went into effect on January 1st, 2001.

I on the other hand is a citizen of a member country of the EU.

#### You failed to get consent
My personal data should **never** have been included in the data set. As **you never asked for my consent**. Article 7 of the mentioned [directive](http://eur-lex.europa.eu/LexUriServ/LexUriServ.do?uri=CELEX:31995L0046:en:HTML) makes this clear:

> CRITERIA FOR MAKING DATA PROCESSING LEGITIMATE
>
> Article 7
>
> Member States shall provide that personal data may be processed only if:
>
> (a) the data subject has unambiguously given his consent; or
> (b) processing is necessary for the performance of a contract to which the data subject is party or in order to take steps at the request of the data subject prior to entering into a contract; or
> (c) processing is necessary for compliance with a legal obligation to which the controller is subject; or
> (d) processing is necessary in order to protect the vital interests of the data subject; or
> (e) processing is necessary for the performance of a task carried out in the public interest or in the exercise of official authority vested in the controller or in a third party to whom the data are disclosed; or
> (f) processing is necessary for the purposes of the legitimate interests pursued by the controller or by the third party or parties to whom the data are disclosed, except where such interests are overridden by the interests for fundamental rights and freedoms of the data subject which require protection under Article 1 (1).

None of these mentioned tests hold true in, at least, my case. I’d guess they don’t hold true in many cases at all but that I technically don’t know, maybe you’ve collected consent (a) from everyone else?

#### You’re required to by law
As you failed to make your collection and processing legitimate as per Article 7 **you’re required to by law to remove it**, see Article 12 of the EU [Directive 95/46/EC](http://eur-lex.europa.eu/LexUriServ/LexUriServ.do?uri=CELEX:31995L0046:en:HTML).

> Article 12
>
> Right of access
>
> […]
>
> (b) as appropriate the rectification, **erasure or blocking of data the processing of which does not comply with the provisions of this Directive**, in particular because of the incomplete or inaccurate nature of the data;
> (c) **notification to third parties to whom the data have been disclosed of any rectification, erasure or blocking carried out in compliance with (b)**, unless this proves impossible or involves a disproportionate effort.

This part is often referred to as the _[right to be forgotten](https://en.wikipedia.org/wiki/Right_to_be_forgotten#European_Union)_ in the EU.


_This is at least my understanding of [Directive 95/46/EC](http://eur-lex.europa.eu/LexUriServ/LexUriServ.do?uri=CELEX:31995L0046:en:HTML). However I’m not a lawyer. Although my understanding is that these protections are going to [become even stronger](http://ec.europa.eu/justice/data-protection/files/factsheets/factsheet_data_protection_en.pdf)._

---

**tl;dr** All in all, you’re violating my rights as a citizen of the EU. As such please delete any personal data about me, block it from being added again and if possible notify all third-parties that they may not use the data about me. These requests are backed by [Directive 95/46/EC](http://eur-lex.europa.eu/LexUriServ/LexUriServ.do?uri=CELEX:31995L0046:en:HTML) and [Wet bescherming persoonsgegevens](http://wetten.overheid.nl/BWBR0011468/2010-07-01).

**Comment by cattedoctor** [comment]
The most interesting thing about this is how your behavior and your project are in violation of ACM & IEEE ethical guidelines. Please consider treading more wisely or those who have been harmed may escalate the issue with the ACM & IEEE.

**Comment by madrobby** [comment]
I've reported the `ghtorrent` user as abusive to GitHub, with a link to this thread. The project maintainer clearly doesn't care blatantly stomping on our privacy. I recommend that everyone who wants their email removed does the same.

**Comment by ELLIOTTCABLE** [comment]
(Hey. Let's take this down a notch: the maintainer / creator has made their
point, we've responded to and refuted it in detail … let's give them a
chance to change their mind before pulling out the legal threats, okay?
Just my 2¢.)

**Comment by madrobby** [comment]
@ELLIOTTCABLE how is reporting them for abuse a legal threat? what?

**Comment by gousiosg** [comment]
This got out of hand. The topic of this issue is removing people's emails from the GHTorrent database. I think all opinions, however harsh or abusive, have been heard. This is a short summary of my opinions and actions in response to the issue:

* GHTorrent does not track or process private information or personal data. It only tracks data in the public domain and makes them available in a different, easier to use format. Whatever GHTorrent does, [anybody](http://stackoverflow.com/questions/897586/does-git-publicly-expose-my-e-mail-address) can do ([and](https://gist.github.com/paulmillr/2657075/) [has](http://osrc.dfm.io) [done](https://www.githubarchive.org)).
* I have already complied with all requests for email removals. 
* I have demonstrated that it is pointless to do the above.

The project will create a process to remove email address in a more systematic manner. It will also create a FAQ page with information about how to use the data in a more privacy-respecting manner.

IMO, the real solution to the problem is that GitHub creates a button that will allow people to opt-in or opt-out the usage of their email for such purposes (similar to the "Available for hire" one). In the mean time, people that do not want their email exposed can [keep it private](https://help.github.com/articles/keeping-your-email-address-private).

To the people that I have removed from GHTorrent; please be aware that you will most probably get more emails from researchers (not because of GHTorrent). The following services can be used to get the same information available on GHTorrent, in exactly the same format and in exactly the same volume:

* [GitHub Archive](https://www.githubarchive.org)
* [GitHub Search](https://github.com/search)
* [cirosantilli/all-github-commit-emails](https://github.com/cirosantilli/all-github-commit-emails/)

**Comment by gousiosg** [comment]
An update: access to GHTorrent data has been suspended until we clear up this issue.
